ccl/schemachangerccl: TestBackupSuccess_base_add_column failed

ccl/schemachangerccl.TestBackupSuccess_base_add_column [failed](https://mesolite.cluster.engflow.com/invocations/default/90a8ac15-36dc-4f25-8a83-17aca0699976?testReportRun=1&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NjaGVtYWNoYW5nZXJjY2w6c2NoZW1hY2hhbmdlcmNjbF90ZXN0) on master @ [8c52619671966cf15055f8b2e271ea82b59440c5](https://github.com/cockroachdb/cockroach/commits/8c52619671966cf15055f8b2e271ea82b59440c5):


```
        	            	      │    │    ├── ABSENT                → PUBLIC           IndexName:{DescID: 118 (#118), Name: "tbl_pkey", IndexID: 2 (tbl_pkey+)}
        	            	      │    │    └── VALIDATED             → PUBLIC           ColumnNotNull:{DescID: 118 (#118), ColumnID: 3 (j+), IndexID: 2 (tbl_pkey+)}
        	            	      │    ├── 4 elements transitioning toward TRANSIENT_ABSENT
        	            	      │    │    ├── TRANSIENT_DELETE_ONLY → TRANSIENT_ABSENT TemporaryIndex:{DescID: 118 (#118), IndexID: 3, ConstraintID: 3, SourceIndexID: 1 (tbl_pkey-)}
        	            	      │    │    ├── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 1, IndexID: 3}
        	            	      │    │    ├── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 2, IndexID: 3}
        	            	      │    │    └── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 3 (j+), IndexID: 3}
        	            	      │    ├── 2 elements transitioning toward ABSENT
        	            	      │    │    ├── PUBLIC                → VALIDATED        PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	      │    │    └── PUBLIC                → ABSENT           IndexName:{DescID: 118 (#118), Name: "tbl_pkey", IndexID: 1 (tbl_pkey-)}
        	            	      │    └── 13 Mutation operations
        	            	      │         ├── MakePublicPrimaryIndexWriteOnly {"IndexID":1,"TableID":118}
        	            	      │         ├── SetIndexName {"IndexID":1,"Name":"crdb_internal_in...","TableID":118}
        	            	      │         ├── SetIndexName {"IndexID":2,"Name":"tbl_pkey","TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":1,"IndexID":3,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":2,"IndexID":3,"Kind":2,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":3,"IndexID":3,"Kind":2,"Ordinal":1,"TableID":118}
        	            	      │         ├── MakeValidatedColumnNotNullPublic {"ColumnID":3,"TableID":118}
        	            	      │         ├── MakeValidatedPrimaryIndexPublic {"IndexID":2,"TableID":118}
        	            	      │         ├── MakeIndexAbsent {"IndexID":3,"TableID":118}
        	            	      │         ├── MakeWriteOnlyColumnPublic {"ColumnID":3,"TableID":118}
        	            	      │         ├── RefreshStats {"TableID":118}
        	            	      │         ├── SetJobStateOnDescriptor {"DescriptorID":118}
        	            	      │         └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"PostCommitNonRev..."}
        	            	      ├── Stage 2 of 3 in PostCommitNonRevertiblePhase
        	            	      │    ├── 3 elements transitioning toward ABSENT
        	            	      │    │    ├── PUBLIC    → ABSENT      IndexColumn:{DescID: 118 (#118), ColumnID: 1, IndexID: 1 (tbl_pkey-)}
        	            	      │    │    ├── PUBLIC    → ABSENT      IndexColumn:{DescID: 118 (#118), ColumnID: 2, IndexID: 1 (tbl_pkey-)}
        	            	      │    │    └── VALIDATED → DELETE_ONLY PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	      │    └── 5 Mutation operations
        	            	      │         ├── MakeWriteOnlyIndexDeleteOnly {"IndexID":1,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":1,"IndexID":1,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":2,"IndexID":1,"Kind":2,"TableID":118}
        	            	      │         ├── SetJobStateOnDescriptor {"DescriptorID":118}
        	            	      │         └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"PostCommitNonRev..."}
        	            	      └── Stage 3 of 3 in PostCommitNonRevertiblePhase
        	            	           ├── 1 element transitioning toward TRANSIENT_ABSENT
        	            	           │    └── PUBLIC      → TRANSIENT_ABSENT IndexData:{DescID: 118 (#118), IndexID: 3}
        	            	           ├── 2 elements transitioning toward ABSENT
        	            	           │    ├── DELETE_ONLY → ABSENT           PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	           │    └── PUBLIC      → ABSENT           IndexData:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-)}
        	            	           └── 5 Mutation operations
        	            	                ├── MakeIndexAbsent {"IndexID":1,"TableID":118}
        	            	                ├── CreateGCJobForIndex {"IndexID":1,"TableID":118}
        	            	                ├── CreateGCJobForIndex {"IndexID":3,"TableID":118}
        	            	                ├── RemoveJobStateFromDescriptor {"DescriptorID":118,"JobID":1007973940837253100}
        	            	                └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"all stages compl..."}
        --- FAIL: TestBackupSuccess_base_add_column/post_commit_stage_7_of_7/restore_database (0.95s)
=== RUN   TestBackupSuccess_base_add_column/post_commit_stage_7_of_7
    --- FAIL: TestBackupSuccess_base_add_column/post_commit_stage_7_of_7 (6.95s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupSuccess_base_add_column.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42619

```
=== RUN   TestBackupSuccess_base_add_column/post_commit_stage_7_of_7/restore_database
    backup.go:490: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/backup.go:490
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"CREATE SCHEMA public"}, []string{"CREATE SEQUENCE public.sq1 MINVALUE 1 MAXVALUE 9223372036854775807 INCREMENT 1 START 1"}, []string{"CREATE TABLE public.tbl (i INT8 NOT NULL, k INT8 NULL, j INT8 NOT NULL DEFAULT 42:::INT8, CONSTRAINT tbl_pkey PRIMARY KEY (i ASC))"}}
        	            	actual  : [][]string{[]string{"CREATE SCHEMA public"}, []string{"CREATE SEQUENCE public.sq1 MINVALUE 1 MAXVALUE 9223372036854775807 INCREMENT 1 START 1"}, []string{"CREATE TABLE public.tbl (i INT8 NOT NULL, k INT8 NULL, CONSTRAINT tbl_pkey PRIMARY KEY (i ASC))"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -8,3 +8,3 @@
        	            	  ([]string) (len=1) {
        	            	-  (string) (len=130) "CREATE TABLE public.tbl (i INT8 NOT NULL, k INT8 NULL, j INT8 NOT NULL DEFAULT 42:::INT8, CONSTRAINT tbl_pkey PRIMARY KEY (i ASC))"
        	            	+  (string) (len=95) "CREATE TABLE public.tbl (i INT8 NOT NULL, k INT8 NULL, CONSTRAINT tbl_pkey PRIMARY KEY (i ASC))"
        	            	  }
        	Test:       	TestBackupSuccess_base_add_column/post_commit_stage_7_of_7/restore_database
        	Messages:   	backup contents:
        	            	parent_schema_name,object_name,object_type
        	            	NULL, public, schema
        	            	public, tbl, table
        	            	public, sq1, table
        	            	
        	            	Schema change plan for ALTER TABLE ‹db›.public.‹tbl› ADD COLUMN ‹j› INT8 NOT NULL DEFAULT ‹42›;
        	            	 ├── PostCommitPhase
        	            	 │    └── Stage 1 of 1 in PostCommitPhase
        	            	 │         ├── 2 elements transitioning toward PUBLIC
        	            	 │         │    ├── WRITE_ONLY → VALIDATED PrimaryIndex:{DescID: 118 (#118), IndexID: 2 (tbl_pkey+), ConstraintID: 2, TemporaryIndexID: 3, SourceIndexID: 1 (tbl_pkey-)}
        	            	 │         │    └── WRITE_ONLY → VALIDATED ColumnNotNull:{DescID: 118 (#118), ColumnID: 3 (j+), IndexID: 2 (tbl_pkey+)}
        	            	 │         └── 2 Validation operations
        	            	 │              ├── ValidateIndex {"IndexID":2,"TableID":118}
        	            	 │              └── ValidateColumnNotNull {"ColumnID":3,"IndexIDForValidation":2,"TableID":118}
        	            	 └── PostCommitNonRevertiblePhase
        	            	      ├── Stage 1 of 3 in PostCommitNonRevertiblePhase
        	            	      │    ├── 4 elements transitioning toward PUBLIC
        	            	      │    │    ├── WRITE_ONLY            → PUBLIC           Column:{DescID: 118 (#118), ColumnID: 3 (j+)}
        	            	      │    │    ├── VALIDATED             → PUBLIC           PrimaryIndex:{DescID: 118 (#118), IndexID: 2 (tbl_pkey+), ConstraintID: 2, TemporaryIndexID: 3, SourceIndexID: 1 (tbl_pkey-)}
        	            	      │    │    ├── ABSENT                → PUBLIC           IndexName:{DescID: 118 (#118), Name: "tbl_pkey", IndexID: 2 (tbl_pkey+)}
        	            	      │    │    └── VALIDATED             → PUBLIC           ColumnNotNull:{DescID: 118 (#118), ColumnID: 3 (j+), IndexID: 2 (tbl_pkey+)}
        	            	      │    ├── 4 elements transitioning toward TRANSIENT_ABSENT
        	            	      │    │    ├── TRANSIENT_DELETE_ONLY → TRANSIENT_ABSENT TemporaryIndex:{DescID: 118 (#118), IndexID: 3, ConstraintID: 3, SourceIndexID: 1 (tbl_pkey-)}
        	            	      │    │    ├── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 1, IndexID: 3}
        	            	      │    │    ├── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 2, IndexID: 3}
        	            	      │    │    └── PUBLIC                → TRANSIENT_ABSENT IndexColumn:{DescID: 118 (#118), ColumnID: 3 (j+), IndexID: 3}
        	            	      │    ├── 2 elements transitioning toward ABSENT
        	            	      │    │    ├── PUBLIC                → VALIDATED        PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	      │    │    └── PUBLIC                → ABSENT           IndexName:{DescID: 118 (#118), Name: "tbl_pkey", IndexID: 1 (tbl_pkey-)}
        	            	      │    └── 13 Mutation operations
        	            	      │         ├── MakePublicPrimaryIndexWriteOnly {"IndexID":1,"TableID":118}
        	            	      │         ├── SetIndexName {"IndexID":1,"Name":"crdb_internal_in...","TableID":118}
        	            	      │         ├── SetIndexName {"IndexID":2,"Name":"tbl_pkey","TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":1,"IndexID":3,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":2,"IndexID":3,"Kind":2,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":3,"IndexID":3,"Kind":2,"Ordinal":1,"TableID":118}
        	            	      │         ├── MakeValidatedColumnNotNullPublic {"ColumnID":3,"TableID":118}
        	            	      │         ├── MakeValidatedPrimaryIndexPublic {"IndexID":2,"TableID":118}
        	            	      │         ├── MakeIndexAbsent {"IndexID":3,"TableID":118}
        	            	      │         ├── MakeWriteOnlyColumnPublic {"ColumnID":3,"TableID":118}
        	            	      │         ├── RefreshStats {"TableID":118}
        	            	      │         ├── SetJobStateOnDescriptor {"DescriptorID":118}
        	            	      │         └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"PostCommitNonRev..."}
        	            	      ├── Stage 2 of 3 in PostCommitNonRevertiblePhase
        	            	      │    ├── 3 elements transitioning toward ABSENT
        	            	      │    │    ├── PUBLIC    → ABSENT      IndexColumn:{DescID: 118 (#118), ColumnID: 1, IndexID: 1 (tbl_pkey-)}
        	            	      │    │    ├── PUBLIC    → ABSENT      IndexColumn:{DescID: 118 (#118), ColumnID: 2, IndexID: 1 (tbl_pkey-)}
        	            	      │    │    └── VALIDATED → DELETE_ONLY PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	      │    └── 5 Mutation operations
        	            	      │         ├── MakeWriteOnlyIndexDeleteOnly {"IndexID":1,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":1,"IndexID":1,"TableID":118}
        	            	      │         ├── RemoveColumnFromIndex {"ColumnID":2,"IndexID":1,"Kind":2,"TableID":118}
        	            	      │         ├── SetJobStateOnDescriptor {"DescriptorID":118}
        	            	      │         └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"PostCommitNonRev..."}
        	            	      └── Stage 3 of 3 in PostCommitNonRevertiblePhase
        	            	           ├── 1 element transitioning toward TRANSIENT_ABSENT
        	            	           │    └── PUBLIC      → TRANSIENT_ABSENT IndexData:{DescID: 118 (#118), IndexID: 3}
        	            	           ├── 2 elements transitioning toward ABSENT
        	            	           │    ├── DELETE_ONLY → ABSENT           PrimaryIndex:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-), ConstraintID: 1}
        	            	           │    └── PUBLIC      → ABSENT           IndexData:{DescID: 118 (#118), IndexID: 1 (tbl_pkey-)}
        	            	           └── 5 Mutation operations
        	            	                ├── MakeIndexAbsent {"IndexID":1,"TableID":118}
        	            	                ├── CreateGCJobForIndex {"IndexID":1,"TableID":118}
        	            	                ├── CreateGCJobForIndex {"IndexID":3,"TableID":118}
        	            	                ├── RemoveJobStateFromDescriptor {"DescriptorID":118,"JobID":1007973940837253100}
        	            	                └── UpdateSchemaChangerJob {"IsNonCancelable":true,"JobID":1007973940837253100,"RunningStatus":"all stages compl..."}
        --- FAIL: TestBackupSuccess_base_add_column/post_commit_stage_7_of_7/restore_database (0.95s)
```

--- a/pkg/sql/schemachanger/sctest/framework.go
+++ b/pkg/sql/schemachanger/sctest/framework.go
@@ -949,7 +949,7 @@ func waitForSchemaChangesToFinish(t *testing.T, tdb *sqlutils.SQLRunner) {
 
 func hasLatestSchemaChangeSucceeded(t *testing.T, tdb *sqlutils.SQLRunner) bool {
 	result := tdb.QueryStr(t, fmt.Sprintf(
-		`SELECT status FROM [SHOW JOBS] WHERE job_type IN ('%s') ORDER BY finished DESC LIMIT 1`,
+		`SELECT status FROM [SHOW JOBS] WHERE job_type IN ('%s') ORDER BY finished DESC, job_id DESC LIMIT 1`,
 		jobspb.TypeNewSchemaChange,
 	))
 	return result[0][0] == "succeeded"

